src/lava/lib/optimization/utils/solver_tuner.py

```diff
@@ -21,56 +23,89 @@ def __init__(self, params_grid: ty.Dict):
+            taking as input the current cost and number of steps to solution.
+            This is the function that is maximized by the SolverTuner.
+        fitness_target: float, optional
+            Fitness target to reach. If this is not passed, the full grid is
+            explored before stopping search.
+        seed: int, default=0
+            Seed for randomized grid search.
+
+        Returns
+        -------
+        best_hyperparams: ty.Dict
+            Dictionary containing the hyper-parameters with the highest fitness.
+        success: bool
+            Flag signaling if the fitness_target has been reached. If no
+            fitness_target is passed, the flag is True.
+        """
+        # TODO : Check that hyperparams are arguments for solver
+
+        best_hyperparams = None
+        best_fitness = -float('inf')
+
+        random.Random(seed).shuffle(self._params_grid)
+
+        for params in self._params_grid:
+            np.random.seed(seed)
+            hyperparams = dict(zip(self._params_keys, params))
+            solver_params["hyperparameters"] = hyperparams
+            solver.solve(**solver_params)
             cost = solver.last_run_report["cost"]
             step_to_sol = solver.last_run_report["steps_to_solution"]
-            if cost is not None and cost <= best_cost and (
-                    step_to_sol < best_step_to_sol):
-                best_hyperparameters = hyperparameters
-                best_cost = cost
-                best_step_to_sol = step_to_sol
-                print(f"""Better hyperparameters configuration found!\
-                        \nHyperparameters: {best_hyperparameters}\
-                        \nBest cost: {best_cost}\
-                        \nBest step-to-solution: {best_step_to_sol}""")
-            if stopping_condition is not None and (
-                    stopping_condition(best_cost, best_step_to_sol)):
+            fitness = fitness_fn(cost, step_to_sol)
+            self._store_trial(hyperparams, cost, step_to_sol, fitness)
+            if fitness > best_fitness:
+                best_hyperparams = hyperparams
+                best_fitness = fitness
+                print(f"Better hyperparameters configuration found!\n"
+                      f"Hyperparameters: {best_hyperparams}")
+            if fitness_target is not None and best_fitness >= fitness_target:
                 break
-            # TODO : Add internal logging
 
-        return best_hyperparameters, best_cost, best_step_to_sol
+        success = best_fitness >= (fitness_target or -float('inf'))
+        return best_hyperparams, success
+
+    def _store_trial(self, params, cost, step_to_sol, fitness):
+        new_entry = tuple(map(lambda k: params[k], self._params_keys))
+        new_entry = new_entry + (cost, step_to_sol, fitness)
+        new_entry_array = np.array([new_entry], dtype=self._store_dtype)
+        self._store = np.concatenate([self._store, new_entry_array])

```

Comment:
At the moment, you are iteratively increasing the size of the numpy array. That's resource inefficient and not what numpy arrays have been designed for.
Maybe better to initialize it to be of the numpy array to be of the same size as the number of combinations of hyperparameters. And then fill up the array.

Reply:
Agree that this is not the best thing performance-wise. On the other hand, we don't know how long the SolverTuner will run, so it is possible that most of the rows would remain empty. Is ok with you?

Reply:
That will probably be the more performant approach. Then you can still cut the numpy array once when returning the results.